From f63502cd800deb4bd71febb754511c3847ef7dfc Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Wed, 6 Apr 2022 12:55:13 +0530
Subject: [PATCH] ath12k: add parse of transmit power envelope element
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The transmit power envelope element has some fields for power,
ath12k should parse it according to IEEE Std 802.11ax™‐2021.
This patch is to do it.

Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h |   1 +
 drivers/net/wireless/ath/ath12k/mac.c  | 194 +++++++++++++++++++++++++
 2 files changed, 195 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 7a9d7b8..2961af3 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -245,6 +245,7 @@ struct ath12k_vif {
 	bool rsnie_present;
 	bool wpaie_present;
 	struct ieee80211_chanctx_conf chanctx;
+	struct ath12k_reg_tpc_power_info reg_tpc_info;
 	u32 key_cipher;
 	u8 tx_encap_type;
 	u8 vdev_stats_id;
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index ea141d6..f09980c 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7494,6 +7494,197 @@ static int ath12k_start_vdev_delay(struct ieee80211_hw *hw,
 	return 0;
 }
 
+static u8 ath12k_mac_get_tpe_count(u8 txpwr_intrprt, u8 txpwr_cnt)
+{
+        switch (txpwr_intrprt) {
+        /* Refer "Table 9-276-Meaning of Maximum Transmit Power Count subfield
+         * if the Maximum Transmit Power Interpretation subfield is 0 or 2" of
+         * "IEEE Std 802.11ax 2021".
+         */
+        case IEEE80211_TPE_LOCAL_EIRP:
+        case IEEE80211_TPE_REG_CLIENT_EIRP:
+                txpwr_cnt = txpwr_cnt <= 3 ? txpwr_cnt : 3;
+                txpwr_cnt = txpwr_cnt + 1;
+                break;
+        /* Refer "Table 9-277-Meaning of Maximum Transmit Power Count subfield
+         * if Maximum Transmit Power Interpretation subfield is 1 or 3" of
+         * "IEEE Std 802.11ax 2021".
+         */
+        case IEEE80211_TPE_LOCAL_EIRP_PSD:
+        case IEEE80211_TPE_REG_CLIENT_EIRP_PSD:
+                txpwr_cnt = txpwr_cnt <= 4 ? txpwr_cnt : 4;
+                txpwr_cnt = txpwr_cnt ? (BIT(txpwr_cnt - 1)) : 1;
+                break;
+        }
+
+        return txpwr_cnt;
+}
+
+static u8 ath12k_mac_get_num_pwr_levels(struct cfg80211_chan_def *chan_def)
+{
+        u8 num_pwr_levels;
+
+        if (chan_def->chan->flags & IEEE80211_CHAN_PSD) {
+                switch (chan_def->width) {
+                case NL80211_CHAN_WIDTH_20:
+                        num_pwr_levels = 1;
+                        break;
+                case NL80211_CHAN_WIDTH_40:
+                        num_pwr_levels = 2;
+                        break;
+                case NL80211_CHAN_WIDTH_80:
+                        num_pwr_levels = 4;
+                        break;
+                case NL80211_CHAN_WIDTH_80P80:
+                case NL80211_CHAN_WIDTH_160:
+                        num_pwr_levels = 8;
+                        break;
+                default:
+                        return 1;
+                }
+        } else {
+                switch (chan_def->width) {
+                case NL80211_CHAN_WIDTH_20:
+                        num_pwr_levels = 1;
+                        break;
+                case NL80211_CHAN_WIDTH_40:
+                        num_pwr_levels = 2;
+                        break;
+                case NL80211_CHAN_WIDTH_80:
+                        num_pwr_levels = 3;
+                        break;
+                case NL80211_CHAN_WIDTH_80P80:
+                case NL80211_CHAN_WIDTH_160:
+                        num_pwr_levels = 4;
+                        break;
+                default:
+                        return 1;
+                }
+        }
+
+        return num_pwr_levels;
+}
+
+static void ath12k_mac_parse_tx_pwr_env(struct ath12k *ar,
+                                        struct ieee80211_vif *vif,
+                                        struct ieee80211_chanctx_conf *ctx)
+{
+        struct ath12k_base *ab = ar->ab;
+        struct ath12k_vif *arvif = (void *)vif->drv_priv;
+        struct ieee80211_bss_conf *bss_conf = &vif->bss_conf;
+        struct ieee80211_tx_pwr_env *single_tpe;
+        enum wmi_reg_6g_client_type client_type;
+        int i;
+        u8 pwr_count, pwr_interpret, pwr_category;
+        u8 psd_index = 0, non_psd_index = 0, local_tpe_count = 0, reg_tpe_count = 0;
+        bool use_local_tpe, non_psd_set = false, psd_set = false;
+
+        client_type = WMI_REG_DEFAULT_CLIENT;
+
+        for (i = 0; i < bss_conf->tx_pwr_env_num; i++) {
+                single_tpe = &bss_conf->tx_pwr_env[i];
+                pwr_category = u8_get_bits(single_tpe->tx_power_info,
+                                           IEEE80211_TX_PWR_ENV_INFO_CATEGORY);
+                pwr_interpret = u8_get_bits(single_tpe->tx_power_info,
+                                            IEEE80211_TX_PWR_ENV_INFO_INTERPRET);
+
+                if (pwr_category == client_type) {
+                        if (pwr_interpret == IEEE80211_TPE_LOCAL_EIRP ||
+                            pwr_interpret == IEEE80211_TPE_LOCAL_EIRP_PSD)
+                                local_tpe_count++;
+                        else if (pwr_interpret == IEEE80211_TPE_REG_CLIENT_EIRP ||
+                                 pwr_interpret == IEEE80211_TPE_REG_CLIENT_EIRP_PSD)
+                                reg_tpe_count++;
+                }
+        }
+
+        if (!reg_tpe_count && !local_tpe_count) {
+                ath12k_warn(ab,
+                            "no transmit power envelope match client power type %d\n",
+                            client_type);
+                return;
+        } else if (!reg_tpe_count) {
+                use_local_tpe = true;
+        } else {
+                use_local_tpe = false;
+        }
+        for (i = 0; i < bss_conf->tx_pwr_env_num; i++) {
+                single_tpe = &bss_conf->tx_pwr_env[i];
+                pwr_category = u8_get_bits(single_tpe->tx_power_info,
+                                           IEEE80211_TX_PWR_ENV_INFO_CATEGORY);
+                pwr_interpret = u8_get_bits(single_tpe->tx_power_info,
+                                            IEEE80211_TX_PWR_ENV_INFO_INTERPRET);
+
+                if (pwr_category != client_type)
+                        continue;
+
+                /* get local transmit power envelope */
+                if (use_local_tpe) {
+                        if (pwr_interpret == IEEE80211_TPE_LOCAL_EIRP) {
+                                non_psd_index = i;
+                                non_psd_set = true;
+                        } else if (pwr_interpret == IEEE80211_TPE_LOCAL_EIRP_PSD) {
+                                psd_index = i;
+                                psd_set = true;
+                        }
+                /* get regulatory transmit power envelope */
+                } else {
+                        if (pwr_interpret == IEEE80211_TPE_REG_CLIENT_EIRP) {
+                                non_psd_index = i;
+                                non_psd_set = true;
+                        } else if (pwr_interpret == IEEE80211_TPE_REG_CLIENT_EIRP_PSD) {
+                                psd_index = i;
+                                psd_set = true;
+                        }
+                }
+        }
+
+        if (non_psd_set && !psd_set) {
+                single_tpe = &bss_conf->tx_pwr_env[non_psd_index];
+                pwr_count = u8_get_bits(single_tpe->tx_power_info,
+         	                        IEEE80211_TX_PWR_ENV_INFO_COUNT);
+                pwr_interpret = u8_get_bits(single_tpe->tx_power_info,
+                                            IEEE80211_TX_PWR_ENV_INFO_INTERPRET);
+                arvif->reg_tpc_info.is_psd_power = false;
+                arvif->reg_tpc_info.eirp_power = 0;
+
+                arvif->reg_tpc_info.num_pwr_levels =
+                        ath12k_mac_get_tpe_count(pwr_interpret, pwr_count);
+                for (i = 0; i < arvif->reg_tpc_info.num_pwr_levels; i++) {
+                        ath12k_dbg(ab, ATH12K_DBG_MAC,
+                                   "non PSD power[%d] : %d\n",
+                                   i, single_tpe->tx_power[i]);
+                        arvif->reg_tpc_info.tpe[i] = single_tpe->tx_power[i] / 2;
+                }
+        }
+        if (psd_set) {
+                single_tpe = &bss_conf->tx_pwr_env[psd_index];
+                pwr_count = u8_get_bits(single_tpe->tx_power_info,
+                                        IEEE80211_TX_PWR_ENV_INFO_COUNT);
+                pwr_interpret = u8_get_bits(single_tpe->tx_power_info,
+                                            IEEE80211_TX_PWR_ENV_INFO_INTERPRET);
+                arvif->reg_tpc_info.is_psd_power = true;
+
+                if (pwr_count == 0) {
+                        ath12k_dbg(ab, ATH12K_DBG_MAC,
+                                   "TPE PSD power : %d\n", single_tpe->tx_power[0]);
+                        arvif->reg_tpc_info.num_pwr_levels =
+                                ath12k_mac_get_num_pwr_levels(&ctx->def);
+                        for (i = 0; i < arvif->reg_tpc_info.num_pwr_levels; i++)
+                                arvif->reg_tpc_info.tpe[i] = single_tpe->tx_power[0] / 2;
+                } else {
+                        arvif->reg_tpc_info.num_pwr_levels =
+                                ath12k_mac_get_tpe_count(pwr_interpret, pwr_count);
+                        for (i = 0; i < arvif->reg_tpc_info.num_pwr_levels; i++) {
+                                ath12k_dbg(ab, ATH12K_DBG_MAC,
+                                           "TPE PSD power[%d] : %d\n",
+                                           i, single_tpe->tx_power[i]);
+                                arvif->reg_tpc_info.tpe[i] = single_tpe->tx_power[i] / 2;
+                        }
+                }
+        }
+}
+
 static int
 ath12k_mac_op_assign_vif_chanctx(struct ieee80211_hw *hw,
 				 struct ieee80211_vif *vif,
@@ -7521,6 +7712,9 @@ ath12k_mac_op_assign_vif_chanctx(struct ieee80211_hw *hw,
                            power_type);
                 if (power_type == IEEE80211_REG_UNSET_AP)
                         power_type = IEEE80211_REG_LPI_AP;
+
+				if (arvif->vdev_type == WMI_VDEV_TYPE_STA)
+                        ath12k_mac_parse_tx_pwr_env(ar, vif, ctx);
         }
 
 	/* for some targets bss peer must be created before vdev_start */
-- 
2.17.1

